spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/metrics/OutputMetricsUpdater.scala

```diff
@@ -0,0 +1,73 @@
+package com.datastax.spark.connector.metrics
+
+import java.util.concurrent.{Semaphore, TimeUnit}
+
+import com.datastax.spark.connector.writer.RichStatement
+import org.apache.spark.executor.{DataWriteMethod, OutputMetrics}
+import org.apache.spark.metrics.CassandraConnectorSource
+import org.apache.spark.{SparkEnv, TaskContext}
+
+private[connector] trait OutputMetricsUpdater {
+  def batchSucceeded(stmt: RichStatement, submissionTimestamp: Long, executionTimestamp: Long)
+
+  def batchFailed(stmt: RichStatement, submissionTimestamp: Long, executionTimestamp: Long)
+
+  def finish(): Long
+}
+
+private class DetailedOutputMetricsUpdater(outputMetrics: OutputMetrics) extends OutputMetricsUpdater {
+  private val mutex = new Semaphore(1)
+  private val taskTimer = CassandraConnectorSource.writeTaskTimer.time()
+
+  def batchSucceeded(stmt: RichStatement, submissionTimestamp: Long, executionTimestamp: Long): Unit = {
+    val t = System.nanoTime()
+    CassandraConnectorSource.writeBatchTimer.update(t - executionTimestamp, TimeUnit.NANOSECONDS)
+    CassandraConnectorSource.writeBatchWaitTimer.update(executionTimestamp - submissionTimestamp, TimeUnit.NANOSECONDS)
+    CassandraConnectorSource.writeRowMeter.mark(stmt.rowsCount)
+    CassandraConnectorSource.writeByteMeter.mark(stmt.bytesCount)
+    CassandraConnectorSource.writeSuccessCounter.inc()
+    mutex.acquire()
+    outputMetrics.bytesWritten += stmt.bytesCount
```

Comment:
I suppose that this isn't atomic because this is a part of task metrics, which should regard a single task. This means - one thread. However, what we do in our writer - we use multiple threads to write - that is why we need additional synchronisation.